doc/man/fedfs-lookup-replication.8
.\"@(#)fedfs-lookup-replication.8"
.\"
.\" @file doc/man/fedfs-lookup-replication.8
.\" @brief man page for fedfs-lookup-replication client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH FEDFS-LOOKUP-REPLICATION 8 "@publication-date@"
.SH NAME
fedfs-lookup-replication \- send a FEDFS_LOOKUP_REPLICATION ADMIN protocol request
.SH SYNOPSIS
.B fedfs-lookup-replication
.RB [ \-?d ]
.RB [ \-n
.IR nettype ]
.RB [ \-h
.IR hostname ]
.RB [ \-t
.IR resolvetype ]
.I path
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
FedFS-enabled file servers allow remote administrative access via an
authenticated RPC protocol known as the
.IR "FedFS ADMIN protocol" .
Using this protocol, FedFS administrators manage
FedFS junctions and NSDB connection parameter information
on remote FedFS-enabled file servers.
.SH DESCRIPTION
The
.BR fedfs-lookup-replication (8)
command is part of a collection of low-level single-use programs
that is intended for testing the FedFS ADMIN protocol or for use in scripts.
It sends a single FEDFS_LOOKUP_REPLICATION request to a remote
FedFS ADMIN protocol service.
.P
The FEDFS_LOOKUP_REPLICATION request causes a remote server
to reveal the contents of a replication marker,
or to report cached or immediate NSDB lookup results as that server sees them.
The
.BR fedfs-lookup-replication (8)
command takes a single positional parameter which is the
pathname on the remote server of the junction to be looked up.
The pathname is relative to the root
of the local file system on the remote server.
.P
Resolving a junction means performing an NSDB query with the contents
of the junction to obtain a list of fileset locations, or FSLs,
matching the stored UUID.
The meaning of these is described in more detail in
.BR fedfs (7).
.P
There are three distinct types of replication lookup:
.TP
.B none
The remote server reports the actual contents of the replication stored
on its local disk.
This includes a UUID and the name and port of an NSDB.
If the
.B \-t
option is not specified, this type of lookup is performed.
.TP
.B cache
The remote server reports lookup results it may have cached from
previous replication lookup requests.
This includes a UUID, the name and port of an NSDB, and the cached
list of fileset locations matching the UUID in the replication marker.
Not all FedFS ADMIN service implementations support this type of request.
.TP
.B nsdb
The remote server performs a fresh replication lookup,
and the results are returned.
This includes an UUID, the name and port of an NSDB,
and a list of fileset locations matching the UUID in the replication marker.
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR fedfs-lookup-replication (8)
version information and a usage message on
.IR stderr .
.IP "\fB\-h, \-\-hostname=\fIhostname\fP"
Specifies the hostname of a remote FedFS ADMIN service.
If this option is not specified, the default value is
.BR localhost .
.IP "\fB\-n, \-\-nettype=\fInettype\fP"
Specifies the transport to use when contacting the remote FedFS ADMIN service.
Typically the
.I nettype
is one of
.B tcp
or
.BR udp .
If this option is not specified, the default value is
.BR netpath .
See
.BR rpc (3t)
for details.
.IP "\fB\-p, \-\-path=\fIpathname\fP"
Specifies the location on the remote server where the target FedFS replication
resides.
This pathname is relative to the remote server's physical root directory,
not the remote server's NFS pseudoroot.
.IP "\fB\-t, \-\-resolvetype=\fItype\fP"
Specifies the desired type of resolution.  Valid values for
.I type
are
.BR 0 ,
.BR none ,
.BR fedfs_resolve_none ,
.BR 1 ,
.BR cache ,
.BR fedfs_resolve_cache ,
.BR 2 ,
.BR nsdb ", or"
.BR fedfs_resolve_nsdb .
The value is not case-sensitive.
If this option is not specified, the default value is
.BR none .
.SH EXIT CODES
The remote administration service returns a value that reflects the
success of the requested operation.
.TP
.B FEDFS_OK
The operation succeeded.
.TP
.B FEDFS_ERR_ACCESS
The caller does not have permission to perform the requested operation
on the remote server.
.TP
.B FEDFS_ERR_BADCHAR
The pathname contains a character which is not
supported by the remote server.
.TP
.B FEDFS_ERR_BADNAME
A component of the pathname consisted of valid UTF-8 characters
supported by the remote server,
but the name is not supported by the remote server
as a valid name for the current operation.
.TP
.B FEDFS_ERR_NAMETOOLONG
The length of the pathname exceeds the remote server’s implementation limit.
.TP
.B FEDFS_ERR_LOOP
Too many symbolic links were encountered in resolving the pathname.
.TP
.B FEDFS_ERR_BADXDR
The remote server encountered an XDR decoding error while
processing the request.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_IO
A hard error occurred on the remote server.
.TP
.B FEDFS_ERR_NOTJUNCT
The pathname does not end in a replication.
.TP
.B FEDFS_ERR_NOTLOCAL
The pathname contains a junction in any position other than the last component.
.TP
.B FEDFS_ERR_PERM
The operation was not allowed because the caller is
either not a privileged user or not the owner of an object that
would be modified by the operation.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred on the remote server.
.TP
.B FEDFS_ERR_NSDB_ROUTE
The remote server was unable to find a route to the NSDB.
.TP
.B FEDFS_ERR_NSDB_DOWN
The remote server determined that the NSDB was down.
.TP
.B FEDFS_ERR_NSDB_CONN
The remote server was unable to establish a connection with the NSDB.
.TP
.B FEDFS_ERR_NSDB_AUTH
The remote server was unable to authenticate
and establish a secure connection with the NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP
An LDAP error occurred on the connection between the remote server and NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_VAL
Indicates the same error as FEDFS_ERR_NSDB_LDAP,
and allows an LDAP protocol error value to be returned to the client.
.TP
.B FEDFS_ERR_NSDB_NONCE
The remote server was unable to locate the NCE in the appropriate NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSN
The remote server was unable to locate the given FSN in the appropriate NSDB.
.TP
.B FEDFS_ERR_NSDB_NOFSL
The remote server was unable to locate any FSLs for the given FSN
in the appropriate NSDB.
.TP
.B FEDFS_ERR_NSDB_RESPONSE
The remote server received a malformed response from the NSDB.
This includes situations when an NSDB entry (e.g. FSN or FSL)
is missing a required attribute.
.TP
.B FEDFS_ERR_NSDB_FAULT
An unanticipated error related to the NSDB occurred.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The remote server does not have any connection
parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL
The remote server received an LDAP referral that it was unable to follow.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_VAL
Indicates the same error as FEDFS_ERR_NSDB_LDAP_REFERRAL,
and allows an LDAP protocol error value to be returned back to the client.
.TP
.B FEDFS_ERR_NSDB_LDAP_REFERRAL_NOTFOLLOWED
The remote server received an LDAP referral that it chose not to follow,
either because the remote server does not support following LDAP referrals
or LDAP referral following is disabled.
.TP
.B FEDFS_ERR_NSDB_PARAMS_LDAP_REFERRAL
The remote server received an LDAP referral that it chose not to follow
because the remote server had no NSDB parameters for the NSDB
targeted by the LDAP referral.
.TP
.B FEDFS_ERR_PATH_TYPE_UNSUPP
The remote server does not support the specified FedFsPathType value.
.TP
.B FEDFS_ERR_NOTSUPP
The remote server does not support the specified procedure.
.TP
.B FEDFS_ERR_NO_CACHE
The remote server does not implement an FSN-to-FSL cache.
.TP
.B FEDFS_ERR_UNKNOWN_CACHE
The software receiving the ONC RPC request is unaware if the remote server
implements an FSN-to-FSL cache or unable to communicate with the
local FSN-to-FSL cache if it exists.
.TP
.B FEDFS_ERR_NO_CACHE_UPDATE
The remote server was unable to update its FSN-to-FSL cache.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that your domain's NSDB hostname is
.IR nsdb.example.net .
You have created a FedFS replication on file server
.IR fs.example.net .
To see how the replication appears on the remote server, use:
.RS
.sp
$ fedfs-lookup-replication -h fs.example.net /export/replication1
.br
Call completed successfully
.br
FSN UUID: 89c6d208-7280-11e0-9f1d-000c297fd679
.br
NSDB: nsdb.example.net:389
.sp
.RE
To see real-time replication resolution results as the remote server sees them, use:
.RS
.sp
$ fedfs-lookup-replication -h fs.example.net -t nsdb /export/replication1
.br
Server returned FEDFS_ERR_NSDB_NOFSN
.sp
.RE
In this example, the replication marker exists on the file server,
but the domain's NSDB has not yet been updated to contain a list of
fileset locations for the UUID contained in the replication marker.
The file server is therefore not able to resolve the replication.
.SH SECURITY
RPCSEC GSSAPI authentication has not yet been implemented for this command.
.SH "SEE ALSO"
.BR fedfs (7),
.BR rpc.fedfsd (8),
.BR rpc (3t)
.sp
RFC 5716 for FedFS requirements and overview
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/fedfsc/fedfs-lookup-replication.c
/**
 * @file src/fedfsc/fedfs-lookup-replication.c
 * @brief Send a FEDFS_LOOKUP_REPLICATION RPC to a FedFS ADMIN server
 */

/*
 * Copyright 2010 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>

#include <rpc/clnt.h>
#include <uuid/uuid.h>
#include <ldap.h>

#include "fedfs.h"
#include "fedfs_admin.h"
#include "nsdb.h"
#include "junction.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Default RPC request timeout
 */
static struct timeval fedfs_lookup_replication_timeout = { 25, 0 };

/**
 * Short form command line options
 */
static const char fedfs_lookup_replication_opts[] = "?dh:n:t:";

/**
 * Long form command line options
 */
static const struct option fedfs_lookup_replication_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "hostname", 1, NULL, 'h', },
	{ "nettype", 1, NULL, 'n', },
	{ "resolvetype", 1, NULL, 't', },
	{ NULL, 0, NULL, 0, },
};

static void
fedfs_lookup_replication_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [-d] [-n nettype] [-h hostname] "
			"[-t <none|cache|nsdb>] path\n\n", progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-n, --nettype        RPC transport (default: 'netpath')\n");
	fprintf(stderr, "\t-h, --hostname       ADMIN server hostname (default: 'localhost')\n");
	fprintf(stderr, "\t-t, --resolvetype    Type of desired result (default: 'none')\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	exit((int)FEDFS_ERR_INVAL);
}

static _Bool
fedfs_lookup_replication_get_resolvetype(const char *resolvetype, FedFsResolveType *resolve)
{
	if (strcmp(resolvetype, "0") == 0 ||
	    strcasecmp(resolvetype, "none") == 0 ||
	    strcasecmp(resolvetype, "fedfs_resolve_none") == 0) {
		*resolve = FEDFS_RESOLVE_NONE;
		return true;
	}
	if (strcmp(resolvetype, "1") == 0 ||
	    strcasecmp(resolvetype, "cache") == 0 ||
	    strcasecmp(resolvetype, "fedfs_resolve_cache") == 0) {
		*resolve = FEDFS_RESOLVE_CACHE;
		return true;
	}
	if (strcmp(resolvetype, "2") == 0 ||
	    strcasecmp(resolvetype, "nsdb") == 0 ||
	    strcasecmp(resolvetype, "fedfs_resolve_nsdb") == 0) {
		*resolve = FEDFS_RESOLVE_NSDB;
		return true;
	}
	return false;
}

static void
fedfs_lookup_replication_print_uuid(const char *pre_text, const FedFsUuid uuid)
{
	char buf[FEDFS_UUID_STRLEN];
	uuid_t uu;

	memcpy(uu, uuid, sizeof(uu));
	uuid_unparse(uu, buf);
	printf("%s: %s\n", pre_text, buf);
}

static void
fedfs_lookup_replication_print_nsdbname(const char *pre_text,
		const FedFsNsdbName nsdbname)
{
	if (nsdbname.hostname.utf8string_val == NULL) {
		printf("%s: NSDB name was empty\n", pre_text);
		return;
	}
	printf("%s: %.*s:%u\n", pre_text,
		nsdbname.hostname.utf8string_len,
		nsdbname.hostname.utf8string_val,
		nsdbname.port);
}

static void
fedfs_lookup_replication_print_fsn(const FedFsFsn fsn)
{
	fedfs_lookup_replication_print_uuid("Fsn UUID", fsn.fsnUuid);
	fedfs_lookup_replication_print_nsdbname("Fsn NSDB name", fsn.nsdbName);
}

static void
fedfs_lookup_replication_print_nfs_fsl(FedFsNfsFsl fsl)
{
	FedFsStatus status;
	char **path_array;
	unsigned int i;

	fedfs_lookup_replication_print_uuid("Fsl UUID", fsl.fslUuid);
	if (fsl.hostname.utf8string_val == NULL)
		printf("FSL hostname: empty\n");
	else
		printf("FSL hostname: %.*s:%u\n",
			fsl.hostname.utf8string_len,
			fsl.hostname.utf8string_val,
			fsl.port);
	status = nsdb_fedfspathname_to_path_array(fsl.path, &path_array);
	if (status != FEDFS_OK)
		printf("Returned NFS export pathname was invalid: %s\n",
			nsdb_display_fedfsstatus(status));
	else {
		if (path_array[0] == NULL)
			printf(" FSL NFS pathname: /\n");
		else {
			printf(" FSL NFS pathname: ");
			for (i = 0; path_array[i] != NULL; i++)
				printf("/%s", path_array[i]);
			printf("\n");
		}

		nsdb_free_string_array(path_array);
	}
}

static void
fedfs_lookup_replication_print_fsl(FedFsFsl fsl)
{
	switch (fsl.type) {
	case FEDFS_NFS_FSL:
		fedfs_lookup_replication_print_nfs_fsl(fsl.FedFsFsl_u. nfsFsl);
		break;
	default:
		printf("Unsupported FSL type\n");
	}
}

static void
fedfs_lookup_replication_print_resok(FedFsLookupResOk result)
{
	unsigned int i;

	fedfs_lookup_replication_print_fsn(result.fsn);

	if (result.fsl.fsl_len == 0) {
		printf("Empty FSL list\n");
		return;
	}
	printf("Returned FSLs:\n");
	for (i = 0; i <= result.fsl.fsl_len; i++)
		fedfs_lookup_replication_print_fsl(result.fsl.fsl_val[i]);
}

static void
fedfs_lookup_replication_print_ldapresultcode(FedFsLookupRes result)
{
	int ldap_result = result.FedFsLookupRes_u.ldapResultCode;
	printf("LDAP result code (%d): %s\n",
				ldap_result, ldap_err2string(ldap_result));
}

static void
fedfs_lookup_replication_print_result(FedFsLookupRes result)
{
	nsdb_print_fedfsstatus(result.status);
	switch (result.status) {
	case FEDFS_OK:
		fedfs_lookup_replication_print_resok(result.FedFsLookupRes_u.resok);
		break;
	case FEDFS_ERR_NSDB_LDAP_VAL:
		fedfs_lookup_replication_print_ldapresultcode(result);
		break;
	default:
		break;
	}
}

static int
fedfs_lookup_replication_call(const char *hostname, const char *nettype,
		const char *path, const char *resolvetype)
{
	FedFsLookupRes result;
	enum clnt_stat status;
	FedFsLookupArgs arg;
	char **path_array;
	CLIENT *client;

	memset(&arg, 0, sizeof(arg));

	if (!fedfs_lookup_replication_get_resolvetype(resolvetype, &arg.resolve))
		return FEDFS_ERR_INVAL;
	arg.path.type = FEDFS_PATH_SYS;
	result.status = nsdb_posix_to_path_array(path, &path_array);
	if (result.status != FEDFS_OK) {
		fprintf(stderr, "Failed to encode pathname: %s",
			nsdb_display_fedfsstatus(result.status));
		return result.status;
	}
	result.status = nsdb_path_array_to_fedfspathname(path_array,
						&arg.path.FedFsPath_u.adminPath);
	if (result.status != FEDFS_OK) {
		fprintf(stderr, "Failed to encode pathname: %s",
			nsdb_display_fedfsstatus(result.status));
		nsdb_free_string_array(path_array);
		return result.status;
	}

	client = clnt_create(hostname, FEDFS_PROG, FEDFS_V1, nettype);
	if (client == NULL) {
		clnt_pcreateerror("Failed to create FEDFS client");
		result.status = FEDFS_ERR_SVRFAULT;
		goto out;
	}

	memset((char *)&result, 0, sizeof(result));
	status = clnt_call(client, FEDFS_LOOKUP_REPLICATION,
				(xdrproc_t)xdr_FedFsLookupArgs, (caddr_t)&arg,
				(xdrproc_t)xdr_FedFsLookupRes, (caddr_t)&result,
				fedfs_lookup_replication_timeout);
	if (status != RPC_SUCCESS) {
		clnt_perror(client, "FEDFS_LOOKUP_REPLICATION call failed");
		result.status = FEDFS_ERR_SVRFAULT;
	} else {
		fedfs_lookup_replication_print_result(result);
		clnt_freeres(client,
			(xdrproc_t)xdr_FedFsLookupRes,
			(caddr_t)&result);
	}
	(void)clnt_destroy(client);

out:
	nsdb_free_fedfspathname(&arg.path.FedFsPath_u.adminPath);
	nsdb_free_string_array(path_array);
	return result.status;
}

int
main(int argc, char **argv)
{
	char *progname, *hostname, *nettype, *path, *resolvetype;
	unsigned int seconds;
	FedFsStatus status;
	int arg;

	(void)setlocale(LC_ALL, "");
	(void)umask(S_IRWXO);

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	hostname = "localhost";
	nettype = "netpath";
	resolvetype = "none";
	path = NULL;
	while ((arg = getopt_long(argc, argv, fedfs_lookup_replication_opts,
				fedfs_lookup_replication_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'h':
			hostname = optarg;
			break;
		case 'n':
			nettype = optarg;
			break;
		case 'p':
			path = optarg;
			break;
		case 't':
			resolvetype = optarg;
			break;
		default:
			fprintf(stderr, "Invalid command line argument: %c\n", (char)arg);
		case '?':
			fedfs_lookup_replication_usage(progname);
		}
	}
	if (argc == optind + 1)
		path = argv[optind];
	else if (argc > optind + 1) {
		fprintf(stderr, "Unrecognized positional parameters\n");
		fedfs_lookup_replication_usage(progname);
	} else {
		fprintf(stderr, "No replication pathname was specified\n");
		fedfs_lookup_replication_usage(progname);
	}

	for (seconds = FEDFS_DELAY_MIN_SECS;; seconds = fedfs_delay(seconds)) {
		status = fedfs_lookup_replication_call(hostname, nettype,
							path, resolvetype);
		if (status != FEDFS_ERR_DELAY)
			break;

		xlog(D_GENERAL, "Delaying %u seconds...", seconds);
		if (sleep(seconds) != 0)
			break;
	}
	return (int)status;
}
